### ISSUE TITLE
MI DB reload commands do not failover properly

### ISSUE THREAD

**Issue by clifjones** [OpenSIPS/opensips#1626]
Hi!
I was testing OpenSIPS 2.4 DB failover and noticed that the MI commands that cause a DB table reload are not timing out as expected. I have proxy configured for a primary and secondary PostgreSQL host using the db_virtual module, which seems to be detecting the DB failure quickly.

Here is the OpenSIPS version and relevant version info:

```
# opensips -V
version: opensips 2.4.4 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 52ff74af8
main.c compiled on 13:37:39 Mar  4 2019 with gcc 7

loadmodule "db_postgres.so"
modparam("db_postgres", "exec_query_threshold", 500000)
modparam("db_postgres", "timeout", 5)

loadmodule "db_virtual.so"
modparam("db_virtual", "db_urls", "define set1 ROUND")
modparam("db_virtual", "db_urls", "postgres://user:pw@primary-host.localnet/postgres")
modparam("db_virtual", "db_urls", "postgres://user:pw@replica-host.localnet/postgres")
modparam("db_virtual", "db_max_consec_retrys", 5)
```
git_version: "52ff74af8702a2830ebcbba3870ccedeea9b261d"

I start the test with both DB connections reachable and process some traffic first.
Then, I block the primary DB connection as follows:

```
# iptables -A OUTPUT -d 10.10.10.10 -j DROP
```

Here are the 2 reload commands that I was testing and their times during the degraded DB state:

```
# time opensipsctl dr reload

real	1m0.307s
user	0m0.021s
sys	0m0.005s
# time opensipsctl dispatcher reload

real	0m20.321s
user	0m0.015s
sys	0m0.012s
```

Each time these commands are called during a DB outage, this type of timeout occurs.

When DB (primary-host) is not-blocked or after it is unblocked, reload time goes to normal:

```
# time opensipsctl dr reload;time opensipsctl dispatcher reload

real	0m0.042s
user	0m0.020s
sys	0m0.007s

real	0m0.025s
user	0m0.025s
sys	0m0.001s
```

It appears that the MI reload commands are not using the db_virtual module's failover logic.

Thanks for your help!

**Comment by bcnewlin** [comment]
Not only are they not failing over per db_virtual, they are also not honoring the db_postgres timeout value setting of 5s.

**Comment by bogdan-iancu** [comment]
The reload commands are simply using the DB backend you configured in the corresponding module - they are not aware or responsible for anything related to the DB failover. This is solely responsibility  of the db_virtual module.

if you run (after running the reloads) a [db_get](https://opensips.org/html/docs/modules/2.4.x/db_virtual.html#mi_db_get) MI command (from db_virtual), what is the status of the real DB conns ?

**Comment by clifjones** [comment]
Unfortunately, the failover issue is even worse than what I originally reported.
I had hoped that commit 280a4dafca390b980a686882783c53bd81b44201, which closed #1443 helped our observed PostgreSQL hangs, but it didn't.

Changing the above test to use the following network outage to block INCOMING primary DB server packets instead of blocking the OUTGOING has dramatic effects.

`iptables -A INPUT -s 10.10.10.10 -j DROP
`

The original block outgoing DB server packets outage would at least result in the DB timeouts showing up close to the set db_postgres timeout period.  In the block incoming db server case, we see no related DB failure logs for 15 minutes.  Then we see a flood of DB timeout, disconnect, failover logs and the db_virtual module will fail over to the replica server full-time (using round-robin selection).
I'm attaching a trap for OpenSIPS 2.4 at commit 60097425d0de465fbc3866b13146b851935e26e6, with my db_virtual failover log patch and the same DB config, etc. as shown the opening post above.

Before the simulated outage, during the 15-minute hang and after the replica failover (still in primary outage), The following command showed both DB connections as GOOD. :-(

```
# opensipsctl fifo db_get
SET::  i=0 name=set1 mode=ROUND
	DB::  j=0 name=postgres://user:pw@primary-host.localnet/postgres can=1 may=1 r_rec=0
	DB::  j=1 name=postgres://user:pw@replica-host.localnet/postgres can=1 may=1 r_rec=0
	
*** MI command showing both DB connections as good.

```
[gdb_opensips_stuck_db_conn.txt](https://github.com/OpenSIPS/opensips/files/3290679/gdb_opensips_stuck_db_conn.txt)

Please let me know if you need additional information.  We are still investigating this ourselves, hoping to find the root cause.

**Comment by bogdan-iancu** [comment]
@clifjones , the trap shows the MI (and other) processes stuck in performing the Postgres query (in waiting for the answer). So, the top issue is the very long long timeout on the Postgres query -> this means the db_virtual fails to detect the failure on the DB failure and to switch to other db conn.

**Comment by clifjones** [comment]
I'm starting to believe that there are similar issues with db_virtual when doing normal call processing.
I use the db_get command as well as my logs patched into the ```dbase.c:db_generic_operation2``` and find that failover is not working as expected during route queries that are in the config files.
I commented on this stale issue that implies this same behavior: #1482

**Comment by bogdan-iancu** [comment]
@clifjones , again, IMO the fault is relater to the long postgres timeout - because of this timeout, it take quite long to OpenSIPS to detect the connection as faulty and to switch to the next connection in the virtual pool.
DId 280a4da have any effect on your setup ? after killing/blocking the prostgres server, how long it takes OpenSIPS to report a new query attempt as failed (due timeout) - does it obey the configured timeout ?

**Comment by clifjones** [comment]
@bogdan-iancu , we are still running failure mode tests with this and finding that certain scenarios seem to work with the 280a4da change, while others result in the calling thread being hung for 15 minutes, despite the timeout value being set to 5 seconds.  This affects call processing and the MI commands when DB reload commands start stacking up.  We are strongly considering alternatives to the db_postgres driver, as it appears to be an issue in libpq and something that may not improve in OpenSIPS without some sort of watchdog on the stuck threads.

**Comment by bogdan-iancu** [comment]
@clifjones , if you manage the isolate the scenarios not affected by the 5 seconds timeout, it will help us to check if the libpq has some options to address them (or not).
If you have the time and resource, it will be great to run your test against other db backend (mysql) just to confirm (or not) if the actual db_virtual does a proper failover.

**Comment by clifjones** [comment]
@bogdan-iancu , we managed to reproduce this issue after removing db_virtual module but still have sql_cacher.  We saw the 
```
ERROR:db_postgres:db_postgres_submit_query: 0x7ff9d12e17a8 PQsendQuery Error: timeout expired
```
logs, that should indicate that a timeout occurred but we still had a lockup in call processing.
We found one serious issue that is most likely causing some or maybe all of this.
The module sql_cacher is leaking read/write locks in many places.  The mi_reload command for that module does not release the write lock on a DB error.  That is just one example of several that we quickly found.  You have to restart the proxy to fix it.

**Comment by bcnewlin** [comment]
In addition to numerous leaked locks in the sql_cacher module, the pain point we are currently running into is that on any reload, from MI or timer, sql_cacher obtains a write_lock on the cache prior to querying the external DB. This blocks all processes from using the cache until the external query returns. In the case where the DB has failed this is very bad as it can block all of OpenSIPS processes until the query times out.

We had implemented sql_cacher to try to remove direct DB access from our scripts due to the timeout issues we were seeing with postgres, but at least with direct DB calls each process can only lock its own connection. With this locking sql_cacher is actually worse as just one reload command can block every call process against the DB query!

**Comment by bogdan-iancu** [comment]
@clifjones & @bcnewlin  - are you guys working together ? Just for me, to know if we have 2 different reports or a single report :)

Now, the initial topic of this ticket is the ability of db_virtual to failover (in combination with db_postgres). So, I would suggest to spin-off as a new ticket the problems you spotted on `sql_cacher` module's locking.

Focusing on this ticket topic, have you manage to get db_virtual properly failover'ing with other backend than postgres ?

**Comment by bcnewlin** [comment]
Sorry, yes we are working together. I agree that the issues with sql_cacher should be in a new ticket.

**Comment by bcnewlin** [comment]
I have opened #1760 for the sql_cacher issues we have seen.

**Comment by bogdan-iancu** [comment]
@clifjones , were you able to narrow down the problem to the db_postgres timeout, affecting the db_virtual detection and switching ?

**Comment by clifjones** [comment]
@bogdan-iancu No, but I can confirm that it still does not work correctly.  The sql_cacher locking issue has improved overall DB error handling but from day 1, the db_virtual module has some weird behavior around reporting which connection that it is actually using. Seems to always report the 1st connection.  We are using ROUND option, so it should alternate.

**Comment by bogdan-iancu** [comment]
@clifjones , let's focus in this ticket on the failover capability of the db_virtual - for the Round-Robin issue, please open a new ticket.
So, what is the latest status in your testing with for failover ? still using db_postgres ? have you tried other DB backend ?

**Comment by clifjones** [comment]
@bogdan-iancu, we have not been using FAILOVER mode on db_virtual.  Here is our config from above:
```
modparam("db_virtual", "db_urls", "define set1 ROUND")
```
We had this configuration with a primary DB and a secondary (replica) connection (2 connections total) in round-robin mode.

When we performed our primary DB outage tests, the logic at the top of the macro that performs a db_virtual query operation, originally hung on the primary and would not fail over properly.  Since then, we have cached our tables so this is less of an issue.  My recent test using the improved sql_cacher code, indicated that when the primary DB connection was offline, the secondary connection was not used by db_virtual.  I will run some more tests to ensure that we are not seeing the libpq 15-minute hang that we have seen in the past.  I'll report our findings.

**Comment by bogdan-iancu** [comment]
OK, still you are testing a failover (under the ROUND setting). So, have you tried this failover scenario (between 2 DBs) with other backend than postgres? sqlite? mysql?
I'm asking as I'm 98% sure that the problem (failed failover) is because of the db_postgres backend and not because of the db_virtual.

**Comment by clifjones** [comment]
We have not tested with anything other than PostgreSQL.  We are considering moving to MySQL, since it appears to be more commonly used on this project.  It is quite likely, that you are correct about the hanging libpq.

**Comment by bogdan-iancu** [comment]
@clifjones, at this point I think this threads should (in)validate 2 potential issues:
1) the failover of the db_virtual module - if you run tests with a non-postgres backend, so we can confirm or not
2) the timeout issue in postgres - maybe this needs to be open again as a separate ticket.
